solidity/test/LPRewardsTest.js

```diff
@@ -0,0 +1,217 @@
+      expect(wrappedTokenBalance).to.eq.BN(web3.utils.toWei("9000"))
+
+      const totalSupply = await lpRewards.totalSupply()
+      expect(totalSupply).to.eq.BN(web3.utils.toWei("9000"))
+
+      const balanceStaker1 = await lpRewards.balanceOf(staker1)
+      expect(balanceStaker1).to.eq.BN(web3.utils.toWei("4000"))
+
+      const balanceStaker2 = await lpRewards.balanceOf(staker2)
+      expect(balanceStaker2).to.eq.BN(web3.utils.toWei("5000"))
+
+      wrappedTokenStakerBalance1 = await wrappedToken.balanceOf(staker1)
+      // 10,000 - 4,000 = 6,000
+      expect(wrappedTokenStakerBalance1).to.eq.BN(web3.utils.toWei("6000"))
+
+      wrappedTokenStakerBalance2 = await wrappedToken.balanceOf(staker2)
+      // 20,000 - 5,000 = 15,000
+      expect(wrappedTokenStakerBalance2).to.eq.BN(web3.utils.toWei("15000"))
+    })
+  })
+
+  describe("tokens distribution", () => {
+    const precision = 1
+
+    it("should be possible to check earned rewards", async () => {
+      const keepRewards = web3.utils.toWei("10000")
+      const wrappedTokenStakerBallance = web3.utils.toWei("10000")
+
+      await mintAndApproveWrappedTokens(staker1, wrappedTokenStakerBallance)
+
+      await fundAndNotifyLPRewards(lpRewards.address, keepRewards)
+
+      const wrappedTokensToStake = web3.utils.toWei("2000")
+      await lpRewards.stake(wrappedTokensToStake, {from: staker1})
+
+      const periodFinish = (await time.latest()).add(time.duration.days(7))
+      await timeIncreaseTo(periodFinish)
+
+      const actualEarnings = (await lpRewards.earned(staker1)).div(
+        tokenDecimalMultiplier
+      )
+      const expectedEarnings = new BN(10000)
+      expect(actualEarnings).to.gte.BN(expectedEarnings.subn(precision))
+      expect(actualEarnings).to.lte.BN(expectedEarnings.addn(precision))
+
+      // duration = 7 days = 604,800 sec
+      // rewardRate = keep rewards / duration
+      // rewardRate = 10,000 / 604,800 = 0,016534391534391534
+      // expected reward per wrapped token = rewardRate * duration / staked amount
+      // expected reward per wrapped token = 5 KEEP
+      // Contract function rewardPerToken() will return 4999999999999999881 KEEP wei.
+      // Solidity does not have floating numbers and when dividing by 10^18,
+      // precision will not be on point.
+      const expectedRewardPerWrappedToken = new BN(5)
+
+      const actualRewardPerToken = (await lpRewards.rewardPerToken()).div(
+        tokenDecimalMultiplier
+      )
+
+      expect(actualRewardPerToken).to.gte.BN(
+        expectedRewardPerWrappedToken.subn(precision)
+      )
+      expect(actualRewardPerToken).to.lte.BN(
+        expectedRewardPerWrappedToken.addn(precision)
+      )
+    })
+
+    it("should be possible to withdraw rewards after staking wrapped tokens", async () => {
+      const rewardsAmount = new BN(5000)
+      const keepAllocated = rewardsAmount.mul(tokenDecimalMultiplier)
+      const wrappedTokenStakerBallance = web3.utils.toWei("10000")
```

Comment:
Typo: ballance->balance.

Reply:
Fixed

Reply:
We are still having the typo in other places, probably as a result of revert, e.g.:
```
      const wrappedTokenStakerBallance1 = web3.utils
        .toBN(10000)
        .mul(tokenDecimalMultiplier)
      const wrappedTokenStakerBallance2 = web3.utils
        .toBN(20000)
        .mul(tokenDecimalMultiplier)

      await mintAndApproveWrappedTokens(staker1, wrappedTokenStakerBallance1)
      await mintAndApproveWrappedTokens(staker2, wrappedTokenStakerBallance2)
```

Reply:
yes, sorry, done